ballerina/memory.bal

```diff
@@ -75,18 +155,60 @@ public isolated class MessageWindowChatMemory {
         readonly & MemoryChatMessage newMessage = check self.mapToMemoryChatMessage(message);
         lock {
             self.createSessionIfNotExist(sessionId);
-            MemoryChatMessage[] memory = self.sessions.get(sessionId);
-            if memory.length() >= self.size - 1 {
-                _ = memory.shift();
-            }
             if newMessage is MemoryChatSystemMessage {
                 self.systemMessageSessions[sessionId] = newMessage;
                 return;
             }
+
+            MemoryChatMessage[] memory = self.sessions.get(sessionId);
+            int memoryLength = memory.length();
+            ROLE newMessageRole = newMessage.role;
+            int size = self.size;
+            SummarizeOverflowConfig? config = self.summarizeOverflowConfig;
+
+            if memoryLength >= size - 1 {
```

Comment:
This addition makes the function long and too nested IMO. This is also somewhere we'd usually use the early return pattern (and potentially break into functions).